articles/node_lifecycle.md

```diff
@@ -0,0 +1,176 @@
+
+[To discuss: Should there be a callback for when configuration is changed, e.g. altered parameters, new topics, etc.?]
+
+While in this state, the node will not receive any execution time to read topics, perform processing of data, respond to functional service requests, etc.
+
+In the inactive state, any data that arrives on topics will not be read. Whether or not data is queued will depend on the QoS parameters of the topics involved. [In other words, to queue data or not is up to each topic.]
+
+Any service requests to a node in the inactive state will not be answered (to the caller, they will fail immediately).
+
+Entering the `inactive` state from the `created` state requires running the node's `onInitialised` method. The node uses this to set up any resources it must hold throughout its life (irrespective of if it is active or inactive). As examples, such resources may include topic publications and subscriptions, memory that is held continuously, and initialising configuration parameters.
+
+Entering the `inactive` state from the `active` state requires running the node's `onDeactivated` method. See the `active` state description for more information.
+
+Entering the `inactive` state from the `fatal error` state requires running the node's `onReset` method. See the `fatal error` state description for more information.
+
+If an error occurs while in this state that cannot be handled by the node, then the transition to the `fatal error` state is taken.
+
+The node may also exit this state by being destroyed, in which case the `destroy` transition is taken.
+
+Implementation note: Causing the transition to the destroyed state can often be mapped to deletion of the node instance.
+
+### Active state
+
+This is the main state of the node's life cycle. While in this state, the node performs any processing, responds to service requests, reads and processes data, produces output, etc.
+
+Entering the `active` state requires running the node's `onActivated` method. This is used to prepare the node for execution. This may include acquiring resources that are only held while the node is actually active, such as access to hardware. Ideally, no preparation that requires significant time (such as lengthy hardware initialisation) should be performed in this callback.
+
+To shut down a node, it must first be deactivated. This will cause the `onDeactivate` method to be called, allowing the node to clean up any resources it only holds while active.
+
+If an error that cannot be handled by the node/system occurs in this state, the transition to the `fatal error` state is taken.
+
+### Fatal Error state
+
+Known errors that can occur in a node should be considered designed behaviour of the system and so should be handled by the node where the error occurred and, if necessary, the surrounding nodes. The reason for this is that the system knows best how to handle errors that occur in its behaviour.
```

Comment:
@doudou: you can program your own drivers however you like, but let's not dictate those kinds of low-level details so early in ROS 2 system design.

>> hardware resets

>Definitely raise error. If you have devices that randomly reset, seriously, change the device vendor. In other cases, the reset is a strong indication of a problem (and that you might not want to trust the device).

Sometimes resetting the hardware is part of the driver's error recovery process. Why do you prefer for actions like that to be done by code that knows less about the characteristics of the device?

>> a higher-level node that has not yet configured some device address parameters

> The whole point of having a lifecycle is to make the nodes a lot simpler by not having to handle these situations. This particular "missing parameter" issue should never happen: if some parameters are critical to the node's function, they should be set before the Initialize transition.

While simplified configuration is an advantage of having a well-defined life cycle, it is certainly *not* the "whole point" of it. 

Many ROS 1 device drivers support [dynamic_reconfigure](http://wiki.ros.org/dynamic_reconfigure), and that is very convenient for users, who can fine-tune device parameters while continuously observing the results. 

ROS 2 will almost certainly provide similar capabilities, as described in the parameter design document. You may choose not to use them, but I would insist on allowing others to take advantage of them.